### PR TITLE
fix(chat): pin the answer marker position and stop warning on expected no-answer replies (#2121, #2118)

### LINKED ISSUE #2118: Every no-answer chat response is misclassified and emits a false "API format may have changed" warning
## Summary

The backend's **normal** no-answer response — the canned prose apology it returns when the sources don't support the question — carries **no `TailwindDoc`**. `AnswerRow.is_answer` therefore evaluates `False`, the parser falls out of its marked-answer path into the emergency fallback, and emits a WARNING claiming the API format may have changed. The format has not changed: this is the backend's standard no-answer shape, and it occurs on every single no-answer ask.

## Evidence (live-verified, not inferred)

Observed live against the real backend on a no-answer ask:

```
AnswerRow.is_answer -> False
WARNING "No marked answer found; falling back to longest unmarked text (159 chars).
         The API response format may have changed."
```

The warning text is `src/notebooklm/_chat/wire.py:233-234`. The flag it depends on is `AnswerRow.is_answer` (`src/notebooklm/_row_adapters/chat.py:739-751`), which requires the type block to be present and marked; the no-answer response has no `TailwindDoc` at all, so the marker can never be found.

## Why this matters

Two distinct problems, neither of which is data loss — callers still receive the apology text, not an empty string:

1. **Misclassification.** A perfectly well-formed, expected backend response is routed through the parser's emergency path. Any future logic that keys off "did we take the fallback?" (metrics, error handling, retries) will treat routine no-answers as anomalies.
2. **A misleading diagnostic that erodes a real signal.** RPC/wire-format drift is documented in `CLAUDE.md` as this project's #1 breakage class. A warning that says "the API response format may have changed" and fires routinely on the most ordinary of responses trains operators to ignore that exact log string — so when the format genuinely does change, the warning is noise.

## Suggested fix

Recognise the no-answer shape explicitly (no `TailwindDoc` present at all, as distinct from a `TailwindDoc` present but unmarked) and route it through a normal, non-warning path. Reserve the "format may have changed" WARNING for the case where the answer document *is* present but the marker is missing — the only case that actually indicates drift.

---

Source: `docs/notes/web-rpc-vs-mobile-grpc-audit-2026-08-07.md`, §1.13.

### LINKED ISSUE #2121: is_answer reads a negative index into a positional wire message — a new TailwindDoc tag silently breaks every chunk
## Summary

`AnswerRow.is_answer` decides whether a streamed chunk is the answer by reading a **negative index** into a positional protobuf-derived message. It works today only because the field it wants happens to be the last one currently registered. The moment the backend adds one more field, the read silently lands on the wrong value and every chunk is classified as "not an answer".

## Evidence (verified against the disassembled schema)

`src/notebooklm/_row_adapters/chat.py:682` sets the marker position to `-1`:

```python
_ANSWER_MARKER_POS: ClassVar[int] = -1
```

and `is_answer` (`src/notebooklm/_row_adapters/chat.py:739-751`) reads `type_block[_ANSWER_MARKER_POS]`, i.e. `first[4][-1]`.

`first[4]` is `AnswerResponse.responseDoc` = `TailwindDoc`, whose `BuilderInfo` registers:

```
body    = 1
(2, 3)  = unused
objects = 4
type    = 5      <-- what [-1] resolves to
```

So `[-1]` reaches `TailwindDoc.type` **only because tag 5 is the highest registered tag today.** Nothing about the read expresses "I want tag 5"; it expresses "I want whatever is last".

## Why this matters

`CLAUDE.md` names undocumented RPC/wire drift as this project's **#1 breakage class** — the obfuscated method IDs and positional shapes are explicitly called out as the things that break when Google changes them. A negative index into a positional wire message is exactly the latent landmine that class of breakage produces:

- The backend adds `TailwindDoc` tag 6 (a routine, backward-compatible protobuf change that breaks nothing for a tag-addressed client).
- `[-1]` now reads the new field instead of `type`.
- `is_answer` goes `False` for **every** chunk.
- The parser falls through to its "No marked answer found" path on every single ask.

There is no error, no exception, and no obviously wrong output — just a silent, total misclassification, surfaced only by a warning that (per the related no-answer misclassification issue) already fires routinely for benign reasons. This is a one-line change to make the failure mode impossible.

## Suggested fix

Replace `_ANSWER_MARKER_POS = -1` with an explicit `4` (the position for tag 5) plus a length guard, so that an unexpected trailing field is ignored rather than silently read as the marker. If a future field genuinely relocates the marker, an explicit index fails visibly instead of quietly inverting the classification.

---

Source: `docs/notes/web-rpc-vs-mobile-grpc-audit-2026-08-07.md`, §1.18.

### PR DESCRIPTION
**Pull request by audichuang**
## Summary

Two defects in the streamed-chat answer path, both in the same read of the same wire slot, so they
ship together:

1. **#2121 — `is_answer` read a negative index into a positional message.**
   `AnswerRow._ANSWER_MARKER_POS` was `-1`, i.e. "whatever field is last". It landed on
   `TailwindDoc.type` only because tag 5 is the highest tag registered today. One routine,
   backward-compatible protobuf field addition would have made `[-1]` read the new field, flipping
   `is_answer` to `False` for **every** chunk of **every** ask — with no error and no exception.
2. **#2118 — the fallback path warned on a normal answer-less reply.**
   When no chunk is marked as the answer, the parser falls back to the longest unmarked text and logs
   `"the API response format may have changed"` at WARNING. But a backend reply of "the sources do not
   contain enough information" is answer-less *by design*: it carries no `TailwindDoc` at all. So the
   drift warning fired on ordinary, healthy usage.

They compound: #2121's failure mode is silent and its only symptom would be #2118's warning firing on
every ask — a warning that (before this PR) already fired routinely for benign reasons and so trained
the reader to ignore it. `CLAUDE.md` names undocumented wire drift as this project's #1 breakage
class, and this warning is how an operator finds out. Fixing the index without de-noising the warning
would leave the alarm untrustworthy; de-noising without fixing the index would remove the only signal
that the index broke.

## Related Issue

Closes #2121
Closes #2118

## Root cause

**#2121.** `first[4]` is `AnswerResponse.responseDoc` = `TailwindDoc`, whose `BuilderInfo` registers
`body = 1`, `(2, 3)` unused, `objects = 4`, `type = 5`. Nothing in `[-1]` expressed "I want tag 5"; it
expressed "I want whatever is last", and tag 5 merely happens to be last.

**#2118.** The parser had one bucket for "no marked answer", covering two different situations:

| Stream | Meaning | Old behaviour | New behaviour |
|---|---|---|---|
| unmarked text, **no** `TailwindDoc` on the row | expected answer-less reply | warns ❌ | **returns quietly** |
| unmarked text **with** a `TailwindDoc` present | the marker moved / real drift | warns | **warns** (unchanged) |
| `TailwindDoc` present but malformed (e.g. a dict) | drift | warns | **warns** (unchanged) |
| zero parseable chunks | drift or empty body | raises | raises (unchanged) |

## Changes

- `src/notebooklm/_row_adapters/chat.py`
  - `_ANSWER_MARKER_POS: -1 -> 4` (0-indexed position of tag 5), and `is_answer`'s length guard
    becomes `len(type_block) > _ANSWER_MARKER_POS` so an unexpected trailing field is **ignored**
    rather than silently read as the marker. Position docs on the class updated from `first[4][-1]`
    to `first[4][4]`.
  - New `has_response_doc` property: whether the `responseDoc` slot is present and non-`None`.
- `src/notebooklm/_chat/wire.py`
  - `_extract_chunk_with_parseable` returns a 6th element, `has_response_doc`, for the row it selected.
  - `parse_streaming_chat_response` accumulates `saw_unmarked_response_doc |= has_response_doc` across
    unmarked chunks and emits the drift warning only when that flag is set. The chosen answer text and
    refs are unchanged — this changes **only** whether the warning is logged.
  - `extract_answer_and_refs_from_chunk` keeps its 4-tuple public shape and discards the new element.
- `tests/_guardrails/_wire_contract.py` — `AnswerRow._ANSWER_MARKER_POS` moves out of `UNMAPPED`
  (whose note said "negative index (-1) … only lands on TailwindDoc.type because tag 5 is currently
  last — see #2121") and into `MAPPINGS` as `TailwindDoc.type`. The guardrail now asserts the mapping
  rather than recording the hazard.
- `tests/unit/test_chat_row_adapter.py` — position-contract tuple updated to `(0, 2, 4, 4, 3, 1)`, plus
  `test_trailing_type_field_does_not_change_answer_marker`, which appends a trailing field to the type
  block and asserts `is_answer` is still `True`. That test is the whole point of #2121 and fails on the
  old `-1`.
- `tests/unit/test_streaming_chat_wire.py` — the `_chunk` helper gains `response_doc: bool` so a row
  can omit the doc, plus four tests covering every row of the table above:
  - `test_no_answer_without_response_doc_does_not_log_drift_warning` (the #2118 fix)
  - `test_no_answer_with_shortened_row_does_not_log_drift_warning` (row too short to carry the slot)
  - `test_present_unmarked_doc_warns_even_when_absent_doc_text_is_longer` (pins the `|=` accumulation:
    the warning must survive even when the *selected* longest text came from a doc-less row)
  - `test_malformed_present_response_doc_still_logs_drift_warning` (a dict at the slot is drift, and
    `has_response_doc` deliberately reports `True` there even though `_type_block` rejects it as
    non-list — so real drift keeps warning)

No public API, exception type, or return type changes. No architectural shape change, so no ADR.

## Test Plan

- [x] I tested these changes locally
- [x] Tests pass (`uv run pytest --cov=src/notebooklm --cov-report=term-missing --cov-fail-under=90`)
- [x] Linting and formatting pass (`uv run pre-commit run --all-files`)
- [x] Type checking passes (`uv run mypy src/notebooklm --ignore-missing-imports`)
- [x] If this PR changes architectural shape, an ADR has been added or updated — N/A, no shape change

Run against the canonical contributor install **plus the extras the CI test matrix adds**
(`--extra browser --extra dev --extra markdown --extra mcp --extra server --extra impersonate`).
The lean `browser+dev+markdown` install alone skips ~1,500 tests and lands at 83% coverage, i.e.
under the gate — so the suite below is the full-extras run that matches CI:

```console
$ uv run mypy src/notebooklm --ignore-missing-imports
Success: no issues found in 351 source files

$ uv run pre-commit run --all-files
ruff.....................................................................Passed
ruff-format..............................................................Passed

$ uv run pytest --cov=src/notebooklm --cov-report=term-missing --cov-fail-under=90 -q
TOTAL                                                    30939    826   8128    507    96%
Required test coverage of 90% reached. Total coverage: 96.41%
15227 passed, 65 skipped, 3 xfailed, 116 warnings in 583.26s (0:09:43)
```

I also ran every remaining gate from the `Code Quality` job of `.github/workflows/test.yml`, all green:
`check_workflow_permissions`, `check_workflow_secret_gates`, `check_action_pinning`,
`check_deprecation_targets`, `check_coverage_thresholds`, `check_ci_install_parity`,
`check_claude_md_freshness`, `check_docs_module_refs`,
`audit_public_api_compat.py --check-stale`, `tests/scripts/check_method_coverage.py`,
`pytest tests/e2e --collect-only`, and the lean-install core-import assertion
(`import notebooklm, notebooklm.notebooklm_cli, notebooklm.client, notebooklm.artifacts`).

## Notes

**Why one PR for two issues.** Both change the same read of `first[4]` in the same two files, and the
guardrail entry for `_ANSWER_MARKER_POS` moves as part of #2121 while #2118's tests need the
`response_doc=False` helper the same change introduces. Splitting them would put a mechanical conflict
between two PRs for no review benefit, and would ship the de-noised warning without the index fix that
makes the remaining warning meaningful.

**`has_response_doc` intentionally diverges from `_type_block`.** `_type_block` additionally requires
`isinstance(block, list)`; `has_response_doc` tests only presence-and-not-`None`. That is deliberate:
a non-list at the slot is drift, so it must keep warning.
`test_malformed_present_response_doc_still_logs_drift_warning` pins that, and tightening
`has_response_doc` to match `_type_block` would silence a genuine drift signal.

**The #2118 suppression is stream-wide, and I want to flag its limit rather than oversell it.** The flag
is set by *any* unmarked doc-bearing chunk in the stream, not by the chunk whose text ends up winning.
That matters because a normal stream does carry unmarked doc-bearing chunks: `TYPE_THOUGHT`
(`ResponseType 2`, per `docs/mobile/enums.txt:1932-1935`) rows have a `TailwindDoc`, and decoding
`tests/cassettes/chat_ask.yaml` shows two of them at 289 and 394 chars. So a no-answer stream that *also*
carried thoughts would still warn.

I do not think that is the observed shape, and I checked before deciding: #2118's live evidence quotes
the warning as `falling back to longest unmarked text (159 chars)`, and 159 is the apology. Thought text
streams cumulatively (289 → 394 in that cassette, up to 531 in `chat_ask_with_references.yaml`), so had
any thought chunk been present in that no-answer stream, the longest unmarked text would have been the
thought, and the author would have quoted ~400 chars and a thinking-text answer instead. The four
recorded ask cassettes cannot exercise this path at all — every one has a `marker=1` answer row, so the
marked branch wins and neither `main` nor this branch logs anything.

The change is also strictly subtractive: it only ever *suppresses* a warning, and it does not touch
which text wins. I verified that by running the thought-plus-apology case against both trees —
`origin/main` and this branch produce identical output. So the worst case here is "the warning still
fires, exactly as it does today", never a new failure. If you would rather have the tighter version, the
principled form is to key the decision on the *winning* row and treat only an absent-or-unenumerated
`type` value as drift (so `TYPE_THOUGHT` stays quiet); say so and I will add it here with a
thinking-plus-apology fixture.

**Scope of verification.** Offline only: unit tests plus the guardrail suite. The claim that
`TailwindDoc.type` is tag 5 is from the disassembled-schema evidence in #2121 (recorded in
`docs/notes/web-rpc-vs-mobile-grpc-audit-2026-08-07.md` §1.18), not an independent capture — I have no
account positioned to re-verify against a live stream. If tag 5 is ever *relocated* (as opposed to a
new tag being appended), an explicit `4` fails visibly instead of quietly inverting every
classification, which is the intended trade.

**Not verified locally:** the Python 3.10–3.14 matrix. No version-sensitive syntax is used.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved streaming chat parsing for missing, truncated, unmarked, or malformed response metadata.
  * Prevented misleading wire-format warnings when metadata is unavailable.
  * Stabilized answer detection when trailing fields are present.
  * Ensured expected no-answer responses return the appropriate fallback message.

* **Documentation**
  * Updated the changelog with the revised no-answer behavior.

* **Tests**
  * Added coverage for response-metadata edge cases and stable answer detection.

### CODE REVIEW

**Comment by audichuang** [comment]
Submission was premature. The branch and worktree are retained for manual review by the user.

**Comment by teng-lin** [comment]
@codex review
@claude review

File: src/notebooklm/_chat/wire.py [hunk 1]
```diff
@@ -229,11 +234,12 @@ def process_chunk(json_str: str) -> None:
         longest_answer = best_marked_answer
         final_refs = best_marked_refs
     elif best_unmarked_answer:
-        logger.warning(
-            "No marked answer found; falling back to longest unmarked "
-            "text (%d chars). The API response format may have changed.",
-            len(best_unmarked_answer),
-        )
+        if saw_unmarked_response_doc:
```
[the remark this answers is not in the document]
Comment (audichuang) [reply src/notebooklm/_chat/wire.py]: Good catch, and partly fixed in ab744e2f — with the rest deliberately left out and flagged rather than guessed at.

**You're right that the schema has the authoritative signal.** `docs/mobile/schema.proto:483-488`:

```proto
message AnswerResponse {
  string response = 1;
  ConversationTurnKey conversationTurnKey = 3;
  EmptyAnswerReason emptyAnswerReason = 4;
  TailwindDoc responseDoc = 5;
}
```

so 0-indexed `0 / 2 / 3 / 4`, and `EmptyAnswerReason` is `0=EMPTY_ANSWER_REASON_UNKNOWN, 1=UNANSWERABLE, 2=FILTERED`. Position 3 is indeed unread today — worth noting `AnswerRow._CITATIONS_POS = 3` is *inside* the type block (`first[4][3]`), not this slot, so there is no collision.

**Fixed: the truncated-row hole.** That was the real defect and you're right that my own test pinned it as intended. `AnswerRow` gains `_EMPTY_ANSWER_REASON_POS` (mapped in the wire-contract registry, not left `UNMAPPED`) and a `suggests_wire_drift` verdict; `wire.py` ORs that instead of raw `has_response_doc`. A row too short to reach `first[3]` could not have carried a reason, so "no doc" proves nothing about it → it warns again. `test_no_answer_with_shortened_row_does_not_log_drift_warning` is now `test_row_truncated_before_the_reason_slot_still_warns` and asserts the opposite.

**Not done: requiring the reason before suppressing.** Two reasons, both worth your call rather than my guess:

1. **No capture shows the field populated on a real no-answer reply.** Every recorded ask cassette carries a *marked* answer, so none of them exercises this path at all, and #2118's live evidence quotes only the warning text and the 159-char fallback length. If `emptyAnswerReason` turns out to be absent on the real apology reply, requiring it would reinstate the exact bug this PR fixes. Making suppression *conditional* on an unobserved field is the riskier direction.
2. **It does not fit the module budget.** Reading it properly (typed reader + `bool` rejection, since `True` would decode as `UNANSWERABLE`) is ~30 lines in `_row_adapters/chat.py`, which sits at 992 of a 1000-line `MODULE_SIZE_BUDGET`. I was not willing to buy the room by deleting unrelated comments.

If you can confirm a live no-answer reply populates position 3, I'll make it the primary signal and drop the doc heuristic entirely — happy to do it here or as a follow-up.

Full suite green: 15226 passed, 96.58% coverage.

**Comment by audichuang** [comment]
Pushed d4ae74de: @claude's stale-doc finding is real and now fixed.

The module-level position table at `src/notebooklm/_row_adapters/chat.py:30-31` still read ``[4][-1] == 1`` marks an answer record — the exact negative-index contract this PR exists to retire — while the `AnswerRow` class docstring had already been updated to ``[4][4]``. Leaving the old form in the file's own position-contract table (the first thing a reader of this module sees) would have kept pointing the next person at the bug. Now ``[4][4] == 1``, matching `_ANSWER_MARKER_POS = 4`.

Docs-only, no behaviour change. Suite green on the branch: 2035 passed, 3 xfailed across `test_chat_row_adapter.py`, `test_streaming_chat_wire.py`, and `tests/_guardrails`.

The open question from the codex thread is unchanged and still yours to call: I did not make `emptyAnswerReason` (slot 3) the primary suppression signal, because no capture shows it populated on a real no-answer reply and requiring an unobserved field would risk reinstating #2118. Confirm a live no-answer reply populates it and I'll switch to it and drop the doc heuristic.

**Comment by teng-lin** [comment]
**Live probe results — settles the `emptyAnswerReason` thread**

Ran 10 live asks across 3 notebooks (control, off-topic, gibberish, PII-style, unanswerable-by-design, wholly-empty notebook, bare punctuation), tapping the raw wire before `parse_streaming_chat_response`. 128 raw `AnswerResponse` rows total. Census over the final 4 asks (42 rows); the earlier 6 asks (86 rows) had an identical shape:

```
len(first) distribution : {5: 42}      ← never truncated
slot3 emptyAnswerReason : {None}       ← never populated, 100% of rows
len(TailwindDoc)        : {5: 42}
TailwindDoc[4] (type)   : {1: 38, 2: 4}
no-TailwindDoc rows     : 0  (every probe)
```

1. **`emptyAnswerReason` is never emitted**, including in the exact "backend declined" scenarios it exists to mark. This matches the 2026-08-07 web-RPC/mobile-gRPC audit (12 live probes, two transports, two accounts, plus 72 cassette rows → zero). Gating suppression on it would suppress nothing and reinstate #2118 on every no-answer reply. Declining the P2 was the right call.

2. **`_ANSWER_MARKER_POS = 4` confirmed live** — `TailwindDoc` is length 5 in all 42 rows with `type` at index 4, so `[-1]` and `[4]` coincide *today* exactly as this PR argues. The pin is correct and the hardening is real.

3. **The truncated-row guard is safe.** I suspected `len(raw) <= 3` might fire on a genuine no-answer row if the transport truncated trailing nulls; live data refutes that — every row is length 5 with an explicit `null` at slot 3. ab744e2f stands as written.

4. **Separately: #2118 no longer reproduces.** Zero no-`TailwindDoc` rows in any probe — every no-answer-ish reply now returns *marked* with a full `responseDoc`, including the wholly-empty notebook the audit used to confirm the bug. The empty-notebook reply is also branded "Gemini Notebook", so backend behaviour has shifted since 2026-08-07. Worth its own issue; not a blocker here, and the suppression branch remains correct defensive code.

Caveat on (4): 10 probes on a single account, and I deliberately did not run the audit's "mild-abusive" trigger, so a narrower path to the unmarked shape may still exist.

The audit note that underpins (1) was untracked locally, which is why neither reviewer could see the evidence — it is being committed separately so this doesn't get re-litigated.

**Comment by teng-lin** [comment]
Merged — thanks @audichuang, this was well-handled work.

Worth calling out specifically: when the review pushed for gating suppression on `emptyAnswerReason`, you fixed the concrete hole that was actually found (truncated rows warning again in ab744e2f) and then **declined to guess at the rest**, with your reasoning written out — no capture showed the field populated, so making suppression conditional on an unobserved field was the riskier direction. That judgement was exactly right. Ten live probes today put slot 3 at null in all 128 rows, so requiring it would have reinstated #2118 on every no-answer reply.

Also appreciated:

- Not buying room against the 1000-line module budget by deleting unrelated comments. Saying "this doesn't fit, here's what it would cost" is the right move over quietly making space.
- The "Not done" section with both reasons stated plainly, rather than silently dropping the harder half.
- Pinning `_ANSWER_MARKER_POS` to 4 and promoting it from `Unmapped` to a real `Mapping` — the live census confirms `TailwindDoc` is length 5 with `type` at index 4, so the old `[-1]` was correct only by coincidence. Genuine latent-bug removal, not cosmetics.
- `test_trailing_type_field_does_not_change_answer_marker` pins the exact failure the old index would have hit.

The one open item from the probe is unrelated to your change: #2118's unmarked shape no longer reproduces at all, which looks like backend drift since the 2026-08-07 audit. Filing that separately.
